fedscale/core/internal/client_metadata.py

```diff
@@ -0,0 +1,73 @@
+import numpy as np
+
+
+class ClientMetadata:
+    """
+    Contains the server-side metadata for a single client.
+    """
+
+    def __init__(self, hostId, clientId, speed, traces=None):
+        """
+        Initializes the ClientMetadata.
+        :param hostId: id of the executor which is handling this client.
+        :param clientId: id of the client.
+        :param speed: computation and communication speed of the client.
+        :param traces: list of client availability traces.
+        """
+        self.hostId = hostId
+        self.clientId = clientId
+        self.compute_speed = speed['computation']
+        self.bandwidth = speed['communication']
+        self.score = 0
+        self.traces = traces
+        self.behavior_index = 0
+
+    def getScore(self):
+        return self.score
+
+    def registerReward(self, reward):
+        """
+        Registers the rewards of the client
+        :param reward: int
+        """
+        self.score = reward
+
+    def isActive(self, cur_time):
+        """
+        Decides whether the client is active at given cur_time
+        :param cur_time: time in seconds
+        :return: boolean
+        """
+        if self.traces is None:
+            return True
+
+        norm_time = cur_time % self.traces['finish_time']
+
+        if norm_time > self.traces['inactive'][self.behavior_index]:
+            self.behavior_index += 1
+
+        self.behavior_index %= len(self.traces['active'])
+
+        if (self.traces['active'][self.behavior_index] <= norm_time <= self.traces['inactive'][self.behavior_index]):
+            return True
+
+        return False
+
+    def getCompletionTime(self, batch_size, local_steps, upload_size, download_size, mean_seconds_per_sample=0.005, tail_skew=0.6):
+        """
+        Computation latency: compute_speed is the inference latency of models (ms/sample). The calculation assumes
+        that client computation speed is a lognormal distribution (see PAPAPYA / GFL papers), and uses the parameters
+        to sample a client task completion time.
+        Communication latency: communication latency = (pull + push)_update_size/bandwidth;
+        :param batch_size: size of each training batch
+        :param local_steps: number of local client training steps
+        :param upload_size: size of model download (MB)
+        :param download_size: size of model upload (MB)
+        :param mean_seconds_per_sample: mean seconds to process a single training example. This can be adjusted based on
+        on-device benchmarks.
+        :param tail_skew: the skew of the lognormal distribution used to model device speed.
+        :return: dict of computation and communication times for the client's training task.
+        """
+        device_speed = max(0.0001, np.random.lognormal(1, tail_skew, 1)[0])
```

Comment:
I am a bit worried about this computation latency. Perhaps you have benchmarked many so that we end up with this 0.005 and 0.6. We need to think about how to avoid breaking many evaluation results that rely on our conventional trace. Of course, one quick workaround is to make `device_speed = 1.0`, and `mean_seconds_per_sample=self.compute_speed`.

Reply:
What is your recommendation for avoiding breaking evaluation results? I'm thinking one option is to add a param to enable or disable the lognormal calculation.

Reply:
<img width="579" alt="image" src="https://user-images.githubusercontent.com/42296458/205092760-a53fe12d-c046-4e39-89c8-fcd8d4f6e40d.png">

the original trace ⬆️ looks like it follows `lognormal` with severe tail latency. btw, what's the reason of choosing `lognormal`?

Reply:
This is a 1-1 mapping between device and speed from AI-Benchmark from my understanding. In practice, the majority of users are using the recent-gen devices. When calibrated based on number of users of that device type, the tail latency should be much less severe. For example, from our experience, at least half of iOS users are using iPhone 11 or later.

Reply:
> at least half of iOS users are using iPhone 11 or later.

what about the rest iOS users' devices?
I think we can at least make the speed scale the same as before (~1e-1 * batch_size * upload_step) so that the benchmark results don't differ too much

OR replace this with more reasonable/realistic number if possible

Reply:
I can't disclose the numbers, but the point I wanted to make is that the distribution is much less tail-skewed when we factor in that most device owners are using recent gen devices rather than the devices that make up the end tail of the distribution.

Reply:
And yes, we can keep the same speed scale.

Reply:
Yea. I think we should make it optional so that we can merge it without breaking existing results.

Reply:
I've made the default computation the existing function to avoid breaking results, and kept the new function un-referenced for now.